File: bootstrap.rb
#!/usr/bin/env ruby

require 'erb'

$supported_packages = [:bash, :tmux, :git, :emacs, :ssh]
$platform = RUBY_PLATFORM
$linux = 'linux'
$osx = 'darwin'

def stow(package)
  unless $supported_packages.include? package
    puts "No package called #{package}"
    return
  end

  if package == :git
    erb = ERB.new(File.read('git/.gitconfig.erb'))
    email = ($platform.include? $osx) ? 'nrcopty@gmail.com' : 'ncopty@appnexus.com'
    namespace = OpenStruct.new(:git_email => email)
    result = erb.result(namespace.instance_eval { binding })
    File.write('git/.gitconfig', result)
  end

  if package == :ssh and platform.include? $osx
    return
  end

  unless system("stow #{package.to_s}")
    puts "There was an error stowing #{package}."
  end
end

def stow_packages(packages=nil)
  unless packages
    $supported_packages.each do |pkg|
      stow pkg.to_sym
    end
  else
    packages.each do |pkg|
      stow pkg.to_sym
    end
  end
end

if __FILE__ == $0
  stow_packages ARGV
end

File: bootstrap.py
class Package(object):

    osx_packages = []

    def install_osx(self):
        for package in self.osx_packages:
            # brew install package
            pass
        pass

    def install_linux(self):
        pass


class Bash(object):

    def pre_stow(self):
        # back up .bash_profile and .bashrc
        pass


class Tmux(object):

    osx_packages = ['tmux']

class Git(object):

    def pre_stow(self):
        # set up gitconfig templates
        pass


class Emacs(object):

    osx_packages = ['emacs', 'cask']


class Ssh(object):
    pass


class Bootstrap(object):

    supported_packages = {
    }

    def stow_package(self, package):
        pass

    def stow(self, packages=None):

        if not packages:
            # all
            pass
        else:
            for package in packages:
                print 'Stowing ...'
                self.stow_package(package)


if __name__ == '__main__':
    pass
